org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -0,0 +1,1296 @@
+    private X509TrustManager[] trust_managers = null;
+
+    private boolean need_client_auth = false;
+    private boolean want_client_auth = false;
+
+    private SSLEngineResult.HandshakeStatus handshake_state = SSLEngineResult.HandshakeStatus.NOT_HANDSHAKING;
+
+    private SSLCipher[] enabled_ciphers = null;
+    private SSLVersion min_protocol = null;
+    private SSLVersion max_protocol = null;
+
+    private JSSSession session = null;
+    private int unknown_state_count = 0;
+
+    private boolean step_handshake = false;
+
+    private boolean is_outbound_closed = false;
+    private boolean is_inbound_closed = false;
+
+    private SSLException ssl_exception = null;
+    private boolean seen_exception = false;
+
+    public JSSEngine() {
+        super();
+
+        peer_info = "";
+        session = new JSSSession(BUFFER_SIZE);
+        logger.debug("JSSEngine: constructor()");
+    }
+
+    public JSSEngine(String peerHost, int peerPort) {
+        super(peerHost, peerPort);
+
+        peer_info = peerHost + ":" + peerPort;
+        session = new JSSSession(BUFFER_SIZE);
+        session.setPeerHost(peerHost);
+        session.setPeerPort(peerPort);
+        logger.debug("JSSEngine: constructor(" + peerHost + ", " + peerPort + ")");
+    }
+
+    public JSSEngine(String peerHost, int peerPort, PK11Cert localCert, PK11PrivKey localKey) {
+        super(peerHost, peerPort);
+
+        peer_info = peerHost + ":" + peerHost;
+        cert = localCert;
+        key = localKey;
+        session = new JSSSession(BUFFER_SIZE);
+        session.setPeerHost(peerHost);
+        session.setPeerPort(peerPort);
+
+        logger.debug("JSSEngine: constructor(" + peerHost + ", " + peerPort + ", " + localCert + ", " + localKey + ")");
+    }
+
+    private String errorText(int error) {
+        // Convert the given error into a pretty string representation with
+        // as much information as is currently available.
+
+        String error_name = PR.ErrorToName(error);
+        String error_text = PR.GetErrorText();
+
+        if (error_name.isEmpty()) {
+            return "UNKNOWN (" + error + ")";
+        } else if (error_text.isEmpty()) {
+            return error_name + " (" + error + ")";
+        } else {
+            return error_name + " (" + error + "): " + error_text;
+        }
+    }
+
+    private void init() {
+        logger.debug("JSSEngine: init()");
+
+        // Initialize our JSSEngine when we begin to handshake; otherwise,
+        // calls to Set<Option>(...) won't be processed if we call it too
+        // early; some of these need to be applied at initialization.
+
+        // Ensure we don't leak ssl_fd if we're called multiple times.
+        if (ssl_fd != null) {
+            try {
+                PR.Close(ssl_fd);
+                ssl_fd = null;
+            } catch (Exception e) {
+                logger.debug("JSSEngine: init() -- error closing previously open ssl_fd: " + e);
```

Comment:
It wouldn't affect the new FD; at worse it'll leak. It is also not really clear in what scenarios the `ssl_fd` will fail to close -- most probably, it would fail to close due to some internal issue with NSS and/or the host is out of memory if it needs to perform another allocation.